// @sveltejs/kit/dist/chunks/write_tsconfig.js
import fs__default from 'fs';
import path__default from 'path';
import { $ } from '../cli.js';
import { m as mkdirp, p as posixify } from './filesystem.js';

/** @type {Map<string, string>} */
const previous_contents = new Map();

/**
 * @param {string} file
 * @param {string} code
 */
function write_if_changed(file, code) {
	if (code !== previous_contents.get(file)) {
		write(file, code);
	}
}

/**
 * @param {string} file
 * @param {string} code
 */
function write(file, code) {
	previous_contents.set(file, code);
	mkdirp(path__default.dirname(file));
	fs__default.writeFileSync(file, code);
}

/** @param {string} str */
function trim(str) {
	const indentation = /** @type {RegExpExecArray} */ (/\n?(\s*)/.exec(str))[1];
	const pattern = new RegExp(`^${indentation}`, 'gm');
	return str.replace(pattern, '').trim();
}

/** @param {string} file */
const exists = (file) => fs__default.existsSync(file) && file;

/** @param {import('types').ValidatedConfig} config */
function write_tsconfig(config, cwd = process.cwd()) {
	const out = path__default.join(config.kit.outDir, 'tsconfig.json');
	const user_file =
		exists(path__default.resolve(cwd, 'tsconfig.json')) || exists(path__default.resolve(cwd, 'jsconfig.json'));

	if (user_file) validate(config, cwd, out, user_file);

	/** @param {string} file */
	const project_relative = (file) => posixify(path__default.relative('.', file));

	/** @param {string} file */
	const config_relative = (file) => posixify(path__default.relative(config.kit.outDir, file));

	const dirs = new Set([
		project_relative(path__default.dirname(config.kit.files.routes)),
		project_relative(path__default.dirname(config.kit.files.lib))
	]);

	/** @type {string[]} */
	const include = [];
	dirs.forEach((dir) => {
		include.push(config_relative(`${dir}/**/*.js`));
		include.push(config_relative(`${dir}/**/*.ts`));
		include.push(config_relative(`${dir}/**/*.svelte`));
	});

	/** @type {Record<string, string[]>} */
	const paths = {};
	const alias = {
		$lib: project_relative(config.kit.files.lib),
		...config.kit.alias
	};
	for (const [key, value] of Object.entries(alias)) {
		if (fs__default.existsSync(project_relative(value))) {
			paths[key] = [project_relative(value)];
			paths[key + '/*'] = [project_relative(value) + '/*'];
		}
	}

	write_if_changed(
		out,
		JSON.stringify(
			{
				compilerOptions: {
					// generated options
					baseUrl: config_relative('.'),
					paths,
					rootDirs: [config_relative('.'), './types'],

					// essential options
					// svelte-preprocess cannot figure out whether you have a value or a type, so tell TypeScript
					// to enforce using \`import type\` instead of \`import\` for Types.
					importsNotUsedAsValues: 'error',
					// Vite compiles modules one at a time
					isolatedModules: true,
					// TypeScript doesn't know about import usages in the template because it only sees the
					// script of a Svelte file. Therefore preserve all value imports. Requires TS 4.5 or higher.
					preserveValueImports: true,

					// This is required for svelte-kit package to work as expected
					// Can be overwritten
					lib: ['esnext', 'DOM'],
					moduleResolution: 'node',
					module: 'esnext',
					target: 'esnext'
				},
				include,
				exclude: [config_relative('node_modules/**'), './**']
			},
			null,
			'\t'
		)
	);
}

/**
 * @param {import('types').ValidatedConfig} config
 * @param {string} cwd
 * @param {string} out
 * @param {string} user_file
 */
function validate(config, cwd, out, user_file) {
	// we have to eval the file, since it's not parseable as JSON (contains comments)
	const user_tsconfig_json = fs__default.readFileSync(user_file, 'utf-8');
	const user_tsconfig = (0, eval)(`(${user_tsconfig_json})`);

	// we need to check that the user's tsconfig extends the framework config
	const extend = user_tsconfig.extends;
	const extends_framework_config = extend && path__default.resolve(cwd, extend) === out;

	const kind = path__default.basename(user_file);

	if (extends_framework_config) {
		const { paths: user_paths } = user_tsconfig.compilerOptions || {};

		if (user_paths && fs__default.existsSync(config.kit.files.lib)) {
			/** @type {string[]} */
			const lib = user_paths['$lib'] || [];
			/** @type {string[]} */
			const lib_ = user_paths['$lib/*'] || [];

			const missing_lib_paths =
				!lib.some((relative) => path__default.resolve(cwd, relative) === config.kit.files.lib) ||
				!lib_.some(
					(relative) => path__default.resolve(cwd, relative) === path__default.join(config.kit.files.lib, '/*')
				);

			if (missing_lib_paths) {
				console.warn(
					$
						.bold()
						.yellow(`Your compilerOptions.paths in ${kind} should include the following:`)
				);
				const relative = posixify(path__default.relative('.', config.kit.files.lib));
				console.warn(`{\n  "$lib":["${relative}"],\n  "$lib/*":["${relative}/*"]\n}`);
			}
		}
	} else {
		let relative = posixify(path__default.relative('.', out));
		if (!relative.startsWith('./')) relative = './' + relative;

		console.warn(
			$.bold().yellow(`Your ${kind} should extend the configuration generated by SvelteKit:`)
		);
		console.warn(`{\n  "extends": "${relative}"\n}`);
	}
}

export { write as a, write_tsconfig as b, trim as t, write_if_changed as w };
